compiler/ir/backend.wasm/src/org/jetbrains/kotlin/backend/wasm/ir2wasm/DeclarationGenerator.kt

```diff
@@ -0,0 +1,329 @@
+import org.jetbrains.kotlin.ir.util.isInterface
+import org.jetbrains.kotlin.ir.util.parentAsClass
+import org.jetbrains.kotlin.ir.visitors.IrElementVisitorVoid
+import org.jetbrains.kotlin.ir.visitors.acceptVoid
+import org.jetbrains.kotlin.wasm.ir.*
+
+class DeclarationGenerator(val context: WasmModuleCodegenContext) : IrElementVisitorVoid {
+
+    // Shortcuts
+    private val backendContext: WasmBackendContext = context.backendContext
+    private val irBuiltIns: IrBuiltIns = backendContext.irBuiltIns
+
+    override fun visitElement(element: IrElement) {
+        error("Unexpected element of type ${element::class}")
+    }
+
+    override fun visitTypeAlias(declaration: IrTypeAlias) {
+        // Type aliases are not material
+    }
+
+    override fun visitFunction(declaration: IrFunction) {
+        // Inline class constructors are currently empty
+        if (declaration is IrConstructor && backendContext.inlineClassesUtils.isClassInlineLike(declaration.parentAsClass))
+            return
+
+        val importedName = declaration.getWasmImportAnnotation()
+
+        // Generate function type
+        val watName = declaration.fqNameWhenAvailable.toString()
+        val irParameters = declaration.getEffectiveValueParameters()
+        val wasmFunctionType =
+            WasmFunctionType(
+                name = watName,
+                parameterTypes = irParameters.map {
+                    val t = context.transformValueParameterType(it)
+                    if (importedName != null && t is WasmRefNullType) {
+                        WasmEqRef
+                    } else {
+                        t
+                    }
+                },
+                resultTypes = listOfNotNull(
+                    context.transformResultType(declaration.returnType).let {
+                        if (importedName != null && it is WasmRefNullType) WasmEqRef else it
+                    }
+                )
+            )
+        context.defineFunctionType(declaration.symbol, wasmFunctionType)
+
+        val isIntrinsic = declaration.hasWasmReinterpretAnnotation() || declaration.getWasmOpAnnotation() != null
+
+        if (declaration is IrSimpleFunction) {
+            if (declaration.modality == Modality.ABSTRACT) return
+            if (declaration.isFakeOverride) return
+
+            val isVirtual = declaration.isOverridableOrOverrides
+
+            if (isVirtual) {
+                // Register function as virtual, meaning this function
+                // will be stored Wasm table and could be called indirectly.
+                context.registerVirtualFunction(declaration.symbol)
+            }
+
+            if (!isVirtual && isIntrinsic) {
+                // Calls to non-virtual intrinsic functions are replaced with something else.
+                // No need to generate them.
+                return
+            }
+        }
+
+        assert(declaration == declaration.realOverrideTarget) {
```

Comment:
It is more about self-checking logic of declaration generator rather than IR validity. We would have to reconstruct context in a separate verifier and keep it in sync with generator, to do the same check.

Reply:
calling `realOverrideTarget` is not free

Reply:
I don't see any specific context required to be reconstructed for this particular check

Reply:
We had a bunch of early returns before we got this assert